{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For index7: mistake(place-amount) = primer-31,ad-42,green-42,flank_simple-21,flank_strange-0;  \n",
      "reads: 1000, good: 0.91, bad: 0.09\n",
      "\n",
      "For new_: mistake(place-amount) = primer-240,ad-73,green-73,flank_simple-48,flank_strange-0;  \n",
      "reads: 3750, good: 0.90, bad: 0.10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# # Raw FASTQ files preprocessing #\n",
    "#################################\n",
    "# \n",
    "# (Latest) library structure (may be changed):\n",
    "# [Alu primer - 12 bp][Alu sequence - 6 bp][Flank][Adapter 1 - 10 bp][Adapter 2, barcode - 9 bp][Adapter3 - 12 bp]\n",
    "# \n",
    "# R1: alu primer mate\n",
    "# R2: adapter mate\n",
    "# \n",
    "# \n",
    "# Steps:\n",
    "# - separate reads into good and bad, depending on the mistake in the primers, the adapter\n",
    "#                                                                            and wrong elements in flanks\n",
    "#     (as well as keeps barcodes good reads in the file for good R2 and location of mistake in the file bad R1)\n",
    "\n",
    "# Import module (main code)\n",
    "import imp\n",
    "import trimmR\n",
    "imp.reload(trimmR)\n",
    "\n",
    "# Variable parameters\n",
    "#################################\n",
    "# The number of permissible error:\n",
    "mist1 = 1\n",
    "mist2 = 1\n",
    "# Primer, ad1 = Adapter 1, ad2 = Adapter 3 (aka Green)\n",
    "primer = 'GAGCCACCGCGC'\n",
    "ad1 = 'GCGTGCTGCGG'\n",
    "ad2 = 'AGGGCGGT'\n",
    "# Length of barcode\n",
    "barlen = 9\n",
    "# List of wrong elements in flank\n",
    "elem_remove = ['AGCT']\n",
    "# Shift for search\n",
    "shift = 4\n",
    "# Window in flanks for searching primers or adapters\n",
    "search_win = 20\n",
    "\n",
    "\n",
    "# Input FASTQ files folder path\n",
    "inputdir = '~/data'\n",
    "inputdir = 'input'\n",
    "# Output folder for processed FASTQ files\n",
    "outputdir = '~/data/processed'\n",
    "outputdir = 'output'\n",
    "#################################\n",
    "\n",
    "# Main function\n",
    "trimmR.main(inputdir, outputdir, shift,\n",
    " mist1, mist2, primer, ad1, ad2, barlen, elem_remove, search_win)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['new_R1_good.fq', 'new_R2_good.fq'], ['index7.R1_good.fq', 'index7.R2_good.fq']]\n",
      "bwa mem chr1/chr1.fa output/new_R1_good.fq output/new_R2_good.fq > foo1/new_.sam\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 6792 sequences (522981 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 255, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (111, 176, 211)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 411)\n",
      "[M::mem_pestat] mean and std.dev: (167.25, 70.72)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 511)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 6792 reads in 3.316 CPU sec, 3.268 real sec\n",
      "[main] Version: 0.7.13-r1126\n",
      "[main] CMD: bwa mem chr1/chr1.fa output/new_R1_good.fq output/new_R2_good.fq\n",
      "[main] Real time: 11.801 sec; CPU: 4.316 sec\n",
      "\n",
      "bwa mem chr1/chr1.fa output/index7.R1_good.fq output/index7.R2_good.fq > foo1/index7.sam\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 1822 sequences (140292 bp)...\n",
      "[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 59, 0, 0)\n",
      "[M::mem_pestat] skip orientation FF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (94, 125, 176)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 340)\n",
      "[M::mem_pestat] mean and std.dev: (133.46, 56.04)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 422)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] skip orientation RR as there are not enough pairs\n",
      "[M::mem_process_seqs] Processed 1822 reads in 0.636 CPU sec, 0.622 real sec\n",
      "[main] Version: 0.7.13-r1126\n",
      "[main] CMD: bwa mem chr1/chr1.fa output/index7.R1_good.fq output/index7.R2_good.fq\n",
      "[main] Real time: 8.878 sec; CPU: 1.424 sec\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#### # Filtered FASTQ files to SAM files  #\n",
    "#################################\n",
    "# \n",
    "# You can see more options for bwa mem on http://bio-bwa.sourceforge.net/bwa.shtml\n",
    "#\n",
    "# Steps:\n",
    "# - run bwa mem (mapping on human genome) with all paired FASTQ in folder.\n",
    "\n",
    "# Import module (main code)\n",
    "import imp\n",
    "import bwamemR\n",
    "imp.reload(bwamemR)\n",
    "\n",
    "# Variable parameters\n",
    "#################################\n",
    "# Input FASTQ files folder path\n",
    "inputdir = '~/data'\n",
    "inputdir = 'output'\n",
    "# Output folder for processed SAM files\n",
    "outputdir = '~/data/processed'\n",
    "outputdir = 'foo1'\n",
    "# Folder with indexed human genome (by bwa index). Name - name of indexed human genome without extension\n",
    "refway = '~/data/name'\n",
    "refway = 'chr1/chr1.fa'\n",
    "# Main part of bwa mem program. You can add options\n",
    "memline = 'bwa mem'\n",
    "#################################\n",
    "\n",
    "# Main function\n",
    "bwamemR.main(inputdir, refway, outputdir, memline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#################################\n",
    "# SAM files filtering           #\n",
    "#################################\n",
    "# \n",
    "# Steps:\n",
    "# - filtering sam files. Reads must be mapped, paired and not duplicated. GOOD reads write in table with\n",
    "#    columns names:\n",
    "#         ID CHR STRAND START END READ1 READ2 BARCODE ALU CIGAR_R1 CIGAR_R2 MDFLAG_R1 MDFLAG_R2\n",
    "#    and BAD reads write in new sam file.\n",
    "\n",
    "# Import module (main code)\n",
    "import imp\n",
    "import samfilterR\n",
    "imp.reload(samfilterR)\n",
    "\n",
    "# Variable parameters\n",
    "#################################\n",
    "# Input SAM files folder path\n",
    "inputdir = '~/data'\n",
    "inputdir = 'foo1'\n",
    "# Output folder for processed SAM files and Tables\n",
    "outputdir = '~/data/processed'\n",
    "outputdir = 'foo2'\n",
    "#################################\n",
    "\n",
    "# Main function\n",
    "samfilterR.main(inputdir, outputdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done new__table\n",
      "\n",
      "Done index7_table\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#################################\n",
    "# Tables clustering             #\n",
    "#################################\n",
    "# \n",
    "# FOR ONE FILE PER CLUSTERING\n",
    "#\n",
    "# Steps:\n",
    "# - clustering reads from tables by distance. Create 2 new tables from 1 with headers:\n",
    "#       1) CLUSTER_ID CHR STRAND START END READ1_BEST CIGAR_BEST MDFLAG_BEST NUM_BARCODES NUM_READS\n",
    "#       2) CLUSTER_ID ID_LIST BARCODE_LIST ALU_LIST\n",
    "\n",
    "# Import module (main code)\n",
    "import imp\n",
    "import bigtableR\n",
    "imp.reload(bigtableR)\n",
    "\n",
    "# Variable parameters\n",
    "#################################\n",
    "# Input files folder path\n",
    "inputdir = '~/data'\n",
    "inputdir = 'foo2'\n",
    "# Output folder for processed cluster_Tables\n",
    "outputdir = '~/data/processed'\n",
    "outputdir = 'foo3'\n",
    "# Distance (in bp) for clustering reads\n",
    "window = 20\n",
    "#################################\n",
    "\n",
    "# Main function\n",
    "bigtableR.main(inputdir, outputdir, window)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 12}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCT': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGACT': 1}\n",
      "{'CCGGCT': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGTCT': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCTGCC': 7}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCTGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 9}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCAGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 2}\n",
      "{'CTGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCAGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CTGGCC': 1}\n",
      "{'CCAGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCT': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCG': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 8}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CGGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CGAGGT': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCT': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CAGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CTGGCT': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGTAT': 7}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCT': 3}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 7}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 4}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCAGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGCCC': 1}\n",
      "{'CCAGCC': 4}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 2}\n",
      "{'CTGGCC': 2}\n",
      "{'CCGGCG': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 4}\n",
      "{'CCAGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CTGGCC': 6}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCA': 1}\n",
      "{'CCGGCT': 5}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCT': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 8}\n",
      "{'CCGGCT': 6}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 8}\n",
      "{'CCGACC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCAGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGCCA': 5}\n",
      "{'CCGGCC': 2}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCT': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCAGCC': 1}\n",
      "{'CCGGCA': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCAGCC': 8}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCT': 1}\n",
      "{'CCGGCC': 7}\n",
      "{'CCGGCC': 7}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CGGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 9}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 6}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCTGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCA': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 5}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'GCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCT': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGCCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGACC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCA': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'TCGGCC': 4}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CTGGCC': 6}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 7}\n",
      "{'CCGGCC': 1}\n",
      "{'TCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCAGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CTGGCC': 3}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 8}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CTGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 4}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 4}\n",
      "{'CAGGCC': 1}\n",
      "{'CCGGCC': 3}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 2}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n",
      "{'CCGGCC': 1}\n"
     ]
    }
   ],
   "source": [
    "#################################\n",
    "# Tables MEGAclustering         #\n",
    "#################################\n",
    "# \n",
    "# FOR ALL FILES\n",
    "#\n",
    "# Steps:\n",
    "# - clustering reads from tables by distance. Create MEGAtable from all bigtables with headers:\n",
    "#       1) MEGACLUSTER_ID CHR STRAND START END READ1_BEST CIGAR_BEST MDFLAG_BEST + \n",
    "#                           for every file creating 2 additional columns: FILENAME_NUM_BARCODES FILENAME_NUM_READS \n",
    "\n",
    "# Import module (main code)\n",
    "import imp\n",
    "import megatableR\n",
    "imp.reload(megatableR)\n",
    "\n",
    "# Variable parameters\n",
    "#################################\n",
    "# Input files folder path\n",
    "inputdir = '~/data'\n",
    "inputdir = 'foo3'\n",
    "# Output folder for processed megacluster_Table\n",
    "outputdir = '~/data/processed'\n",
    "outputdir = 'foo4'\n",
    "# Distance (in bp) for megaclustering reads\n",
    "window = 20\n",
    "# Standart Alu\n",
    "standart_alu = 'CCGGCC'\n",
    "#################################\n",
    "\n",
    "# Main function\n",
    "megatableR.main(inputdir, outputdir, window, standart_alu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#################################\n",
    "# Intersect with existing libs  #\n",
    "#################################\n",
    "#\n",
    "# Steps:\n",
    "# - intersect reads from megatable by distance with existing libraries (like ALU). Add to MEGAtable additional columns:\n",
    "#       Name_of_library1 Name_of_library2 etc (in each library: if seq from library intersect reads from megatable: \n",
    "#                                               writing Name_of_seq, else: 'NA')\n",
    "\n",
    "# Import module (main code)\n",
    "import imp\n",
    "import intersectionR\n",
    "imp.reload(intersectionR)\n",
    "\n",
    "# Variable parameters\n",
    "#################################\n",
    "# Input files folder path\n",
    "inputdir = '~/data'\n",
    "inputtable = 'foo4/megatable.txt'\n",
    "# Input rep_libraries forlder path\n",
    "inputlibrary = '~/data/rep_libraries'\n",
    "inputlibrary = 'alu_rep'\n",
    "# Output folder for processed megacluster_Table with extracolumns (Name_of_library)\n",
    "outputdir = '~/data/processed'\n",
    "outputdir = 'foo5'\n",
    "# Distance (in bp) for clustering reads\n",
    "window = 20\n",
    "#################################\n",
    "\n",
    "# Main function\n",
    "intersectionR.main(inputtable, inputlibrary, outputdir, window)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['MEGACLUSTER_ID', 'FILENAME', 'READNAME', 'CHR', 'POS', 'STRAND', 'ALU_BEST', 'ALU_AMOUNT', 'ALU_HAMMING', 'READ1_BEST', 'READ2_BEST', 'TLEN', 'CIGAR_BEST', 'MDFLAG_BEST', 'new__NUM_READS', 'new__NUM_BARCODES', 'index7_NUM_READS', 'index7_NUM_BARCODES', 'Alu_hg19']\n",
      "['MEGACLUSTER_ID', 'FILENAME', 'READNAME', 'CHR', 'POS', 'STRAND', 'ALU_BEST', 'ALU_AMOUNT', 'ALU_HAMMING', 'READ1_BEST', 'READ2_BEST', 'TLEN', 'CIGAR_BEST', 'MDFLAG_BEST', 'new__NUM_READS', 'new__NUM_BARCODES', 'index7_NUM_READS', 'index7_NUM_BARCODES', 'Alu_hg19', 'READ1', 'READ2']\n"
     ]
    }
   ],
   "source": [
    "#################################\n",
    "# Recovery reads by FASTQ_good  #\n",
    "#################################\n",
    "#\n",
    "# Steps:\n",
    "# - recovery reads by FASTQ_good by (using cols FILENAME and READNAME\n",
    "\n",
    "# Import module (main code)\n",
    "import imp\n",
    "import readfromfqR\n",
    "imp.reload(readfromfqR)\n",
    "\n",
    "# Variable parameters\n",
    "#################################\n",
    "# Input files folder path\n",
    "inputdir = '~/data'\n",
    "inputtable = 'foo5/megatable_inter_lib.txt'\n",
    "# Input rep_libraries forlder path\n",
    "inputlibrary = '~/data/fq_good_folder'\n",
    "inputlibrary = 'output'\n",
    "# Output folder for processed megacluster_Table with extracolumns (Name_of_library)\n",
    "# Distance (in bp) for clustering reads\n",
    "outputdir = '~/data/processed'\n",
    "outputdir = 'foo5'\n",
    "#################################\n",
    "\n",
    "# Main function\n",
    "readfromfqR.main(inputtable, inputlibrary, outputdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
